src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormatter.cs

```diff
@@ -167,5 +157,190 @@ private TextEdit[] MapEditsToHostDocument(RazorCodeDocument codeDocument, TextEd
+                    changes.Add(change);
+
+                    indentationMap.Add(projectedDocumentIndex, new IndentationMapData()
+                    {
+                        OriginalProjectedDocumentIndex = projectedDocumentIndex,
+                        AnnotationAttachIndex = position + previousMarkerOffset,
+                        MarkerKind = MarkerKind.Trivia,
+                    });
+
+                    // We have added a marker. This means we need to account for the length of the marker in future calculations.
+                    previousMarkerOffset += markerString.Length;
+                }
+                else
+                {
+                    // No marker needed. Let's attach the annotation directly at the given location.
+                    indentationMap.Add(projectedDocumentIndex, new IndentationMapData()
+                    {
+                        OriginalProjectedDocumentIndex = projectedDocumentIndex,
+                        AnnotationAttachIndex = projectedDocumentIndex + previousMarkerOffset,
+                        MarkerKind = MarkerKind.Token,
+                    });
+                }
+            }
+
+            var changedText = context.CSharpSourceText.WithChanges(changes);
+            var syntaxTree = CSharpSyntaxTree.ParseText(changedText, cancellationToken: cancellationToken);
+            return (indentationMap, syntaxTree);
+        }
+
+        private SyntaxNode AttachAnnotations(
+            Dictionary<int, IndentationMapData> indentationMap,
+            IEnumerable<int> projectedDocumentLocations,
+            SyntaxNode root)
+        {
+            foreach (var projectedDocumentIndex in projectedDocumentLocations)
+            {
+                var indentationMapData = indentationMap[projectedDocumentIndex];
+                var annotation = new SyntaxAnnotation(MarkerId, $"{projectedDocumentIndex}");
+
+                if (indentationMapData.MarkerKind == MarkerKind.Trivia)
+                {
+                    var trackingTrivia = root.FindTrivia(indentationMapData.AnnotationAttachIndex, findInsideTrivia: true);
+                    var annotatedTrivia = trackingTrivia.WithAdditionalAnnotations(annotation);
+                    root = root.ReplaceTrivia(trackingTrivia, annotatedTrivia);
+                }
+                else
+                {
+                    var trackingToken = root.FindToken(indentationMapData.AnnotationAttachIndex, findInsideTrivia: true);
+                    var annotatedToken = trackingToken.WithAdditionalAnnotations(annotation);
+                    root = root.ReplaceToken(trackingToken, annotatedToken);
+
+                    // Since a token can span multiple lines, we need to keep track of the offset within the token span.
+                    // We will use this later when determining the exact line within a token.
+                    indentationMapData.CharacterOffset = indentationMapData.AnnotationAttachIndex - trackingToken.SpanStart;
```

Comment:
I'll add to the comment. The example is a multiline string literal,
```C#
    var x = @"
This
|is a
single
token";
```
Since all of this string is a single token, if the caller asks for the indentation at `|`, we would provide 4(indentation at token.spanStart) but we should be providing 0.

Reply:
ooo and a caller might do that if they do range formatting at that location right?

Reply:
Well, range formatting simply calls regular formatting in the current design. But yeah, regular formatting extracts the C# desired indentation for each line that starts in C# context and the example above is part of that.

Reply:
I see alright I think I understand that better then